src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,343 @@
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+}
+
+//! Compute gradient over whole dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& coordinates,
+                                        arma::mat& gradient)
+{
+  gradient.zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = 0; i < dataset.n_rows; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      gradient += (1 - regularization) * OuterProduct(cij);
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate gradient due to triplets.
+        arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        arma::vec cil = dataset.col(i) - dataset.col(impostors(l, i));
+
+        gradient += regularization * (OuterProduct(cij) - OuterProduct(cil));
+      }
+    }
+  }
+}
+
+//! Compute gradient over a batch of data points.
+template<typename MetricType>
+template <typename GradType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& coordinates,
+                                        const size_t begin,
+                                        GradType& gradient,
+                                        const size_t batchSize)
+{
+  gradient.zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = begin; i < begin + batchSize; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      gradient += (1 - regularization) * OuterProduct(cij);
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate gradient due to triplets.
+        arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        arma::vec cil = dataset.col(i) - dataset.col(impostors(l, i));
+
+        gradient += regularization * (OuterProduct(cij) - OuterProduct(cil));
+      }
+    }
+  }
+}
+
+} // namespace lmnn
+} // namespace mlpack
+
+namespace mlpack {
+namespace optimization{
+
+inline void Projection(arma::mat& iterate)
+{
+  // Projection.
+  arma::cx_vec eigval;
+  arma::cx_mat eigvec;
+
+  arma::eig_gen(eigval, eigvec, iterate);
+
+  arma::vec realEigVal = arma::real(eigval);
+  arma::mat realEigVec = arma::real(eigvec);
+
+  arma::uvec ind = arma::find(realEigVal > 0);
+  arma::mat diagEigVal = arma::diagmat(realEigVal);
+  iterate = realEigVec.cols(ind) * diagEigVal.submat(ind, ind) *
+      arma::trans(realEigVec.cols(ind));
+}
+
+// Template specialization for the SGD optimizer.
+template<>
+template<typename DecomposableFunctionType>
+double StandardSGD::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
```

Comment:
I think that this will work, but I'm not sure how `Projection()` will work for L_BFGS.  I suspect that maybe you could use a custom update policy for SGD to get the projection in there (instead of this overload).  I'm not sure of the best way to do it for L-BFGS, I'd have to think about it a bit...

Reply:
Hmm, so it depends on whether or not we plan to go with this approach in the long term.  In the short term, for testing, we can do something *really* ugly and at each `Evaluate()` call, cast the `const arma::mat& coordinates` to an `arma::mat&` then project it.  That can work for testing but it is, in my view, far too much of a hack to actually commit long-term.

Alternately, I could see overloading a few optimizers to handle the projection step and then using a `static_assert()` to ensure that an overloaded optimizer is used, or, modifying all of the optimizers to have a `PostStepPolicy` or something like this that allows some action to be taken after a step.  (That last idea would be a lot of work for just one `FunctionType` though.)